### PR TITLE
Write to log when error is shown in PR detail view

### PR DESCRIPTION
**Pull request by jcansdale**
These commits were originally intended to help track down the 'Couldn't find merge base' issue #1190. This issue hasn't been resolved yet, but I think having the extra logging and exception detail will be useful stand alone.

When the following error is displayed, the exception detail will also be written to the log.
couldnt find merge base

### CODE REVIEW

File: src/GitHub.App/Services/PullRequestService.cs [hunk 1]
```diff
@@ -316,11 +316,6 @@ public IObservable<Unit> SwitchToBranch(ILocalRepositoryModel repository, IPullR
                         pullRequest.Head.Sha,
                         pullRequest.Base.Ref,
                         pullRequest.Head.Ref);
-
```
Comment (jcansdale) [inline src/GitHub.App/Services/PullRequestService.cs]: This exception is now thrown from `GetPullRequestMergeBase` (with more info about what went wrong).

File: src/GitHub.App/Services/PullRequestService.cs [hunk 2]
```diff
@@ -350,7 +345,7 @@ static bool HasPreamble(string file, Encoding encoding)
             {
                 foreach (var b in encoding.GetPreamble())
                 {
-                    if(b != stream.ReadByte())
+                    if (b != stream.ReadByte())
```
Comment (jcansdale) [inline src/GitHub.App/Services/PullRequestService.cs]: Look @shana, a space! ;-)

File: src/GitHub.App/ViewModels/PullRequestDetailViewModel.cs [hunk 3]
```diff
@@ -351,6 +354,7 @@ public override void Initialize(ViewWithData data)
                 .ObserveOn(RxApp.MainThreadScheduler)
                 .Catch<IPullRequestModel, Exception>(ex =>
                 {
+                    log.Error("Error observing GetPullRequest", ex);
```
Comment (jcansdale) [inline src/GitHub.App/ViewModels/PullRequestDetailViewModel.cs]: Add logs when exception is displayed on UI.

File: src/GitHub.App/Services/PullRequestService.cs (hunk 1 above)
Comment (jcansdale) [reply src/GitHub.App/Services/PullRequestService.cs]: It appears we still have an issue when commenting in inline mode. This wasn't the line I commented on! 😭

File: src/GitHub.App/Services/PullRequestService.cs [hunk 4]
```diff
@@ -316,11 +316,6 @@ public IObservable<Unit> SwitchToBranch(ILocalRepositoryModel repository, IPullR
                         pullRequest.Head.Sha,
                         pullRequest.Base.Ref,
                         pullRequest.Head.Ref);
-
-                    if (sha == null)
```
Comment (jcansdale) [inline src/GitHub.App/Services/PullRequestService.cs]: This is the line with `sha == null`. How does it look?

File: src/GitHub.App/Services/PullRequestService.cs [hunk 5]
```diff
@@ -316,11 +316,6 @@ public IObservable<Unit> SwitchToBranch(ILocalRepositoryModel repository, IPullR
...
-                    {
-                        throw new NotFoundException($"Couldn't find merge base between {pullRequest.Base.Sha} and {pullRequest.Head.Sha}.");
```
Comment (jcansdale) [inline src/GitHub.App/Services/PullRequestService.cs]: This exception is now thrown from GetPullRequestMergeBase (with more info about what went wrong).

File: src/GitHub.App/Services/PullRequestService.cs (hunk 2 above)
Comment (shana) [reply src/GitHub.App/Services/PullRequestService.cs]: rejoice

File: src/UnitTests/GitHub.App/Services/GitClientTests.cs [hunk 6]
```diff
@@ -194,7 +194,11 @@ public class TheGetPullRequestMergeBaseMethod : TestBaseClass
             repo.Network.Remotes.Add(null, null).ReturnsForAnyArgs(remote);
             var gitClient = new GitClient(Substitute.For<IGitHubCredentialProvider>());
 
-            await gitClient.GetPullRequestMergeBase(repo, baseUri, headUri, baseSha, headSha, baseRef, headRef);
+            try
+            {
+                await gitClient.GetPullRequestMergeBase(repo, baseUri, headUri, baseSha, headSha, baseRef, headRef);
+            }
+            catch (NotFoundException) { }
```
Comment (shana) [inline src/UnitTests/GitHub.App/Services/GitClientTests.cs]: Ignoring an exception in a test is weird. Is it supposed to throw or not?

File: src/GitHub.App/Services/GitClient.cs [hunk 7]
```diff
@@ -394,7 +394,7 @@ public Task<bool> IsModified(IRepository repository, string path, byte[] content
                 baseCommit = repo.Lookup<Commit>(baseSha);
                 if (baseCommit == null)
                 {
-                    return null;
+                    throw new NotFoundException($"Couldn't find {baseSha} after fetching from {baseCloneUrl}:{baseRef}.");
```
Comment (shana) [inline src/GitHub.App/Services/GitClient.cs]: So I understand why you'd want the details of the shas in the log, but this error message is incredibly cryptic, there is no way anyone is going to know what is going on with an error like this, and this is going to be displayed directly to a user where it's going to be completely useless.

We should have a message for the user, and another with more details for the log. If the log is only showing the message and stacktrace, we can set InnerException with the log details and have the outer exception be for the user.

Comment (shana) [reply src/GitHub.App/Services/GitClient.cs]: Also, are all the callers of this method expecting an exception to be thrown and handling it properly?

File: src/UnitTests/GitHub.App/Services/GitClientTests.cs (hunk 6 above)
Comment (jcansdale) [reply src/UnitTests/GitHub.App/Services/GitClientTests.cs]: It's certainly a bit unusual. The method it's calling would previously return a null when a comment couldn't be fetched/found. It now throws an exception.

The test is to check whether or not `Fetch` was called. It doesn't care about the exception or exception message. There is another test for that.

File: src/GitHub.App/Services/PullRequestService.cs (hunk 5 above)
Comment (jcansdale) [reply src/GitHub.App/Services/PullRequestService.cs]: What happens if an exception is thrown from a `Observable.Defer(async () =>`?

File: src/GitHub.App/Services/GitClient.cs (hunk 7 above)
Comment (jcansdale) [reply src/GitHub.App/Services/GitClient.cs]: Callers of this method previously turned the null into an exception themselves. I've also tested it by making this method always throw to check that nothing bad happens. The exception message is displayed as expected.

Comment (jcansdale) [reply src/GitHub.App/Services/GitClient.cs]: I wonder what the user facing message should say? I suspect that for some reason, fetch doesn't always work or there is some race condition. I guess we could suggest thry try again or maybe refresh. I'll investagate...

File: src/UnitTests/GitHub.App/Services/GitClientTests.cs (hunk 6 above)
Comment (shana) [reply src/UnitTests/GitHub.App/Services/GitClientTests.cs]: Ah ok, in that case we should probably document that with a comment, so we know in a year from now what the assumption here is.

File: src/GitHub.App/Services/GitClient.cs (hunk 7 above)
Comment (shana) [reply src/GitHub.App/Services/GitClient.cs]: Yeah, an indication of what the user action should be in the case of this error would be great!

Comment (jcansdale) [reply src/GitHub.App/Services/GitClient.cs]: Something like this.

image

Or maybe simply, "Please check your network connection and try again."?

Comment (jcansdale) [reply src/GitHub.App/Services/GitClient.cs]: "Couldn't find merge base. Please check your network connection and try again."

I think this is maybe the right balance of revealing what actually went wrong, but giving the user a hint that has a good chance of fixing it.

Comment (jcansdale) [reply src/GitHub.App/Services/GitClient.cs]: Here we go:
image

Comment (shana) [reply src/GitHub.App/Services/GitClient.cs]: So "merge base" may be meaningless to a user, and try again might not be helpful either if the action that the user performed to get here isn't available from this screen.

What is the action that failed here? Opening a file? Showing the PR details? The first part of the message should say what was the action that failed.

What is try again? Refresh? Click on the PR button in the toolbar and then clicking on the PR again? The second part should say what the user can do on this screen to attempt to make it work.

Assuming it was the PR that failed to load, then we should probably say something like

`The Pull Request failed to load. Please check your network connection and click refresh to try again.`

Adjust accordingly for the failed action.

Comment (jcansdale) [reply src/GitHub.App/Services/GitClient.cs]: Okay, I've chained it to variants of your suggestion above.

My concern was that we hide the real issue behind a workaround. I'm still keen to get to the bottom of what's going wrong. My hunch is that it's a race condition in gitlib2sharp. I'll see if I can reproduce the issue with some standalone code... 😄

Comment (shana) [reply src/GitHub.App/Services/GitClient.cs]: Yes, we don't want to hide the issue in a workaround. If you're concerned about people not reporting it, you can add "If this issue persists, please let us know at support@github.com"

Comment (jcansdale) [reply src/GitHub.App/Services/GitClient.cs]: Done. Also confirmed that nested exceptions appear in logs.

**Pull request merged**